doc/source/reference/c-api.array.rst

```diff
@@ -243,7 +243,7 @@ From scratch
 
 .. c:function:: PyObject* PyArray_NewLikeArray( \
         PyArrayObject* prototype, NPY_ORDER order, PyArray_Descr* descr, \
-        int subok)
+        int ndim, npy_intp* dims, int subok)
```

Comment:
looks like this breaks backwards compatibility

Reply:
Indeed -- unfortunately we can't add add parameters to functions in NumPy's C API without breaking API compatibility.

I think we need to make a private internal function to handle `shape` on the C side, which could be used by `empty_like()` and `PyArray_NewLikeArray` internally.

Reply:
I see, thanks for pointing that out @rgommers.

@shoyer could you clarify how we would use such a function from `PyArray_NewLikeArray()`? I can see how I could have a completely new function `PyArray_NewLikeArrayWithShape()` being called from `array_empty_like()`, but that would incur on a lot of code duplication. I can't see how we would handle `PyArray_NewLikeArray()` if it doesn't know about the new shape we're trying to use, other than modifying `*prototype`'s shape before and restore it afterwards, but that feels too hacky.

Reply:
Ah, ok, one alternative would be to move the implementation to a new function, and keep `PyArray_NewLikeArray()` only as a wrapper for backwards compatibility, is that what you mean @shoyer? I think this new function didn't even need to be private, it could be the new `PyArray_NewLikeArrayWithShape()` I proposed.

Reply:
indeed, move the implementation to your new function and keep `PyArray_NewLikeArray` as a wrapper for backwards compat